### PR TITLE
[SPARK-11863][SQL] Unable to resolve order by if it contains mixture of aliases and real columns

### PR DESCRIPTION
**Pull request by cloud-fan**
this is based on https://github.com/apache/spark/pull/9844, with some bug fix and clean up.

The problems is that, normal operator should be resolved based on its child, but `Sort` operator can also be resolved based on its grandchild. So we have 3 rules that can resolve `Sort`: `ResolveReferences`, `ResolveSortReferences`(if grandchild is `Project`) and `ResolveAggregateFunctions`(if grandchild is `Aggregate`).
For example, `select c1 as a , c2 as b from tab group by c1, c2 order by a, c2`, we need to resolve `a` and `c2` for `Sort`. Firstly `a` will be resolved in `ResolveReferences` based on its child, and when we reach `ResolveAggregateFunctions`, we will try to resolve both `a` and `c2` based on its grandchild, but failed because `a` is not a legal aggregate expression.

whoever merge this PR, please give the credit to @dilipbiswal

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus @yhuai

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 1]
```diff
@@ -218,6 +218,24 @@ class AnalysisSuite extends AnalysisTest {
       nullResult,
       udf4)
     // checkUDF(udf4, expected4)
+    }
```
Comment (rxin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: indent

**Comment by marmbrus** [comment]
Thanks, I'm going to merge this to master and 1.6.

**Comment by dilipbiswal** [comment]
@marmbrus @cloud-fan Thank you !!

**Pull request closed without merging**